### ISSUE TITLE
hold_jid in aws submission scripts

### ISSUE THREAD

**Issue by mwilensky768** [EoRImaging/pipeline_scripts#10]
In various places of `run_eppsilon_aws.sh` a `qsub` call is made with the `-hold_jid` flag. This is problematic due to the way that parallelcluster schedules compute instances. The cluster is autoscaled according to jobs that are being held in addition to jobs that are being actively computed. This means there is wasted compute time if the max queue size is not set exactly right and timing is not miraculous. Plan is to instead make a daemon that monitors job progress by checking for outputs on s3, and then submitting when those are made.

**Comment by mwilensky768** [comment]
@rlbyrne I tested this today just to verify the behavior and found it only spun up enough instances (1) to do the jobs that were not being held. The jobs being held have the status `hqw`. Do you have a way of reproducing the issue/do you think this behavior could have changed since you observed it?

**Comment by mwilensky768** [comment]
I should also note that this is using `ondemand` instances rather than `spot` instances. Not sure if that matters.

**Comment by rlbyrne** [comment]
Oh interesting! It's definitely possible that this issue has been fixed in the more recent ParallelCluster releases. It shouldn't matter if the instances are on demand or spot.

**Comment by mwilensky768** [comment]
OK, I may have spoken too soon. I just noticed that at some point 8 jobs could run, which required 2 instances according to my allocation. These both were fully occupied for some time, but then some overlap happened where only 1 instance was required (1 more job being held), and the 2nd instance was not shut down immediately. However, I think if you look at the overall compute time required, it is actually still cheaper than doing it in the non-parallel way.

In other words I think it does not upsize when it sees jobs being held, but it also does not reliably and instantaneously downsize when it could.

That said, currently I only have 1 job in the queue at all and it is not instantaneously terminating the last remaining unnecessary instance, so this may just be a behavior of the scheduler in general.

**Comment by rlbyrne** [comment]
It sounds like more testing is required to figure out exactly how ParallelCluster is handing held jobs. You could imagine testing with dummy jobs and cheap instances: submit one that runs for a while and a second that holds for it and look at the request query in the Auto Scaling Groups section in the console (instead of just waiting to see if instances spin up, that's useful for figuring out what's going on under the hood).

When I was first debugging autoscaling clusters I used scripts like this a lot: https://github.com/rlbyrne/rlb_aws/blob/master/sge_test_script.sh

**Comment by mwilensky768** [comment]
OK, I wrote a very simple wrapper that submits 4 jobs to imitate what a parallelized eppsilon job looks like. The middle two jobs are to wait for the first job to complete, using `-hold_jid`. The last job is to wait for the middle two jobs, also using `-hold_jid`. Each job requests an entire `t3a.small` instance. In the first timeline, each job just sleeps for 10 minutes and prints timestamps to the logs. In the second timeline, each job sleeps for 30 minutes (a more realistic timing for a single eppsilon job in the parallelized format). I used the autoscaling activity gui to get a timestamp on when instances get created/terminated.

**First Timeline**
In this timeline, an instance was already in existence at time of submission, so there was minimal wait time to start the first job. Instances usually take about 10 minutes to launch I notice.

3:26:21 - jobs are submitted
3:36:36 - first job starts
3:46:36 - first job ends
3:46:51 - second job starts, third job now transitions from `hqw` to `qw`
3:47:43 - launching a new instance
3:56:51 - second job ends
3:57:06 - third job starts on _first_ instance. Second instance is just a bit too late it seems.
4:07:06 - third job ends.
4:07:21 - fourth job starts on _first_ instance
4:07:23 - second instance is terminated
4:17:21 - fourth job ends
4:27:25 - first instance terminated

This timeline had ten minutes of wasted cpu time due to timing mismatches, and an additional ten minutes of wasted time at the end when no jobs were in the queue.

**Second Timeline**
No instances were in effect when this timeline started.

4:48:16 - Jobs are submitted
4:48:50 - A new instance is launched
4:58:21 - First job starts on first instance
5:28:21 - First job ends
5:28:36 - Second job starts on first instance, third job transitions from `hqw` to `qw`
5:29:07 - A new instance is launched
5:38:51 - Third job starts on second instance
5:58:36 -  Second job finishes, fourth job is still holding for third
6:08:29 - First instance is set to terminate
6:08:51 - Third job finishes
6:09:06 - Fourth Job Starts
6:39:06 - Fourth Job Ends
6:48:59 - Second instance terminated

We see in both timelines that at no point are instances scheduled ahead of when they are needed, although some are perhaps reserved for longer than they ought to be. Ideally, the first instance would be set to terminate immediately after the second job finished, so there was again ten minutes of overlap waste in this timeline. However, there was also ten minutes of overlap waste at the end of both timelines. I don't know if there is a way to get this shorter, but I think this is evidence that the bulk of this issue is no longer a concern. We are basically at the latency of the autoscaler, and it is not scheduling instances beyond that latency. By this I mean, I have not found a case where an instance that is no longer necessary is actually immediately gotten rid of, so I think this is as good as it is going to get without manually handling the scheduler.

Now let's talk about compute costs.

When I am processing an obs, and want its actual power spectrum out, and need a beam per frequency, I have to run on an instance with greater than 140 GB of memory. I use an r5.8xlarge instance, which is the smallest memory-optimized instance that does the job. I need 19 of its cores to do the job, but I may as well schedule all 32, because, assuming I am running only FHD-eppsilon series jobs, there is nothing that will fit in the remaining 13 cores. eppsilon then gets called after FHD, and makes use of all 32 cores. With this many cores, eppsilon needs about 10.6 minutes to do each cube, and there are 16 cubes. This is just under 3 hours of computation, plus some negligible amount to plot. 

On the other hand, if we use the parallelized wrappers and assign 8 cores per job (more than enough memory, makes the compute less slow), then each cube takes about 22 minutes, but each job does at most two cubes (some do 1). The way it works out in the most wasteful case is that one instance gets used for a total of 1.5 hours, and another instance gets "used" for 0.5 hours, but actually computes nothing. This works out to 2 hours of compute time total of an r5.8xlarge instance, which is less than the just under 3 in the case above. So, I think even including waste, which I think would be hard to get rid of, it is still cheaper to use the parallelized wrapper. In the most ideal case, which will only happen if a lot of jobs are being done at once so that the timing overlap is not so bad, each obs takes one hour of compute time, with some wasted schedule time around the edges like we see here.

Some optimization could be done on instance type that might make the costs more comparable.

What do you think @rlbyrne ?

**Comment by rlbyrne** [comment]
Thank you for checking this! It looks like the problem I identified in the past isn't an issue with the latest ParallelCluster. We're good to hold jobs in the queue!

**Comment by mwilensky768** [comment]
Excellent! On that note, I was using aws-parallelcluster 2.8.0 installed via pip. I think in a future PR I will try to come up with a way to intertwine the parallelized eppsilon jobs with the single obs FHD job script, which as outlined above should be both cheaper and faster. I will now close this issue.
